projects/laji/src/app/shared/pipe/label.pipe.ts

```diff
@@ -18,100 +18,70 @@ type labelType = 'qname'|'fullUri'|'warehouse'|'withKey'|'emptyWhenMissing';
-       (type === 'fullUri' && value.indexOf('http') !== 0)) {
+  transform<T extends string|string[]>(value: T, type?: labelType): T {
+    if (!value || (typeof value !== 'string' && !Array.isArray(value)) || value.length === 0) {
       return value;
     }
+    const key = Array.isArray(value) ? value.join(',') : value;
+
     // if we ask another time for the same key, return the last value
-    if (value === this.lastKey) {
-      return this.value;
+    if (key === this.lastKey) {
+      return this.value as T;
     }
-    // store the query, in case it changes
-    this.lastKey = value;
-
-    // set the value
-    this.updateValue(value, type);
+    this.lastKey = key;
 
-    // if there is a subscription to onLangChange, clean it
-    this._dispose();
-
-    // subscribe to onLangChange event, in case the language changes
-    if (!this.onLangChange) {
-      this.onLangChange = this.translate.onLangChange.subscribe(() => {
-        this.lastKey = null; // we want to make sure it doesn't return the same value until it's been updated
-        this.updateValue(value, type);
+    if (Array.isArray(value)) {
+      from(value).pipe(
+        concatMap(v => this.fetchValue(v, type)),
+        toArray()
+      ).subscribe(v => {
+        this.updateValue(v);
       });
+    } else {
+      this.fetchValue(value, type).subscribe(v => this.updateValue(v));
     }
-    return this.value;
-  }
 
-  ngOnDestroy(): void {
-    this._dispose();
+    return this.value as T;
   }
 
-  /**
-   * Clean any existing subscription to onLangChange events
-   */
-  protected _dispose(): void {
-    if (this.onLangChange) {
-      this.onLangChange.unsubscribe();
-      this.onLangChange = undefined;
-    }
+  private updateValue(value: string|string[]) {
+    this.value = value;
+    this.cdr.markForCheck();
   }
 
-  private _updateValue(key: string, type?: labelType): void {
+  private fetchValue(key: string, type?: labelType): Observable<string> {
     let obs;
     switch (type) {
+      case 'warehouse':
+        obs = this.warehouseService.getOriginalKey(key).pipe(
+          concatMap(res => this.fetchValue(res))
+        );
+        break;
       case 'fullUri':
-        obs = this.triplestoreLabelService.get(IdService.getId(key), this.translate.currentLang);
+        obs = key.indexOf('http') === 0 ?
+          this.triplestoreLabelService.get(IdService.getId(key), this.translate.currentLang) :
+          of(key);
         break;
       case 'withKey':
         obs = this.triplestoreLabelService.get(key, this.translate.currentLang).pipe(
           map(value => value !== key ? (value || '') + ' (' + key + ')' : value));
         break;
       default:
-        obs = this.triplestoreLabelService.get(key, this.translate.currentLang);
+        obs = this.triplestoreLabelService.get(key, this.translate.currentLang).pipe(
+          map((res: string) => res && res !== key ? res : (type === 'emptyWhenMissing' ? '' : key))
```

Comment:
Could we split this to multiple lines? Multiple ternary operators in a single line is a bit hard to read.
E.g. 

```
res && res !== key
  ? res
  : type === 'emptyWhenMissing'
    ? ''
    : key
```

Reply:
sounds good  👍